### PR TITLE
Teams Features

### PR DESCRIPTION
**Pull request by petebryan**
This PR covers a range of features to support customers ingesting and hunting in Teams events.

## Proposed Changes

  - Logic App for ingesting Teams events via Office 365 Management Activity API
  - Parser for normalizing Teams events
  - A series of Hunting queries to find anomalous Teams related activity

### CODE REVIEW

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  - Persistance
+relevantTechniques:
+  - T1136
+query: |
+
+  // If you want to look at user added further than 7 days ago adjust this value
+  let time_ago = 7d;
+  // If you want to change the timeframe of how quickly accounts need to be added and removed change this value
+  let time_delta = 1h;
+  TeamsData
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: Would there be a way to get the ClientIP by joining in AzureAD data? - https://docs.microsoft.com/en-us/microsoft-365/compliance/detailed-properties-in-the-office-365-audit-log?view=o365-worldwide

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  let time_delta = 1h;
+  TeamsData 
+  | where TimeGenerated > ago(time_ago)
+  | where Operation == "MemberAdded"
+  | extend UPN = tostring(parse_json(Members)[0].UPN)
+  | project TimeAdded=TimeGenerated, Operation, UPN, UserWhoAdded = UserId
+  | join (
+  TeamsData 
+  | where TimeGenerated > ago(time_ago)
+  | where Operation == "MemberRemoved"
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: Recommend using soft match for this string - =~ versus ==

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+query: |
+
+  // If you want to look at user added further than 7 days ago adjust this value
+  let time_ago = 7d;
+  // If you want to change the timeframe of how quickly accounts need to be added and removed change this value
+  let time_delta = 1h;
+  TeamsData 
+  | where TimeGenerated > ago(time_ago)
+  | where Operation == "MemberAdded"
+  | extend UPN = tostring(parse_json(Members)[0].UPN)
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: We should bring thru TeamName and Details.TeamGuid value as this helps identify the specific Team.

File: Hunting Queries/TeamsLogs/MultiTeamOwner.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  | summarize dcount(TeamName) by Member
+  | where dcount_TeamName > max_owner_count
+  | project Member);
+  TeamsData
+  | where TimeGenerated > ago(time_window)
+  | where Operation == "MemberRoleChanged"
+  | extend Member = tostring(parse_json(Members)[0].UPN) 
+  | extend NewRole = toint(parse_json(Members)[0].Role) 
+  | where NewRole == 2
+  | where Member in (high_owner_count)
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/MultiTeamOwner.yaml]: Same comments as I had for ExternalUserAddedRemoved.yaml

File: Hunting Queries/TeamsLogs/MultipleTeamsDeletes.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  | where TimeGenerated > ago(time_window)
+  | where Operation == "TeamDeleted"
+  | summarize count() by UserId
+  | where count_ > max_delete
+  | project UserId);
+  TeamsData
+  | where TimeGenerated > ago(time_window)
+  | where Operation == "TeamDeleted"
+  | where UserId in (deleting_users)
+  | project-away AddOnName, Members, Settings
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/MultipleTeamsDeletes.yaml]: Same comments as I had for ExternalUserAddedRemoved.yaml

File: Hunting Queries/TeamsLogs/NewBotAdded.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+tactics:
+  - Persistance
+  - Collection
+relevantTechniques:
+  - T1176
+  - T1119
+query: |
+
+  // If you have more than 30 days worth of Teams data change this value
+  let data_date = 30d;
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/NewBotAdded.yaml]: Generally we don't want to have the default be 30 days, we usually want to go with 14days to support better perf on larger orgs.

File: Hunting Queries/TeamsLogs/NewBotAdded.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  let historical_bots = (
+  TeamsData
+  | where TimeGenerated > ago(data_date)
+  | where isnotempty(AddOnName)
+  | project AddOnName);
+  TeamsData
+  | where TimeGenerated > ago(1d)
+  // Look for add-ins we have never seen before
+  | where AddOnName in (historical_bots)
+  | extend timestamp = OfficeActivity_TimeGenerated, AccountCustomEntity = UserId
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/NewBotAdded.yaml]: There is no OfficeActivity_TimeGenerated, change to timestamp = TimeGenerated.

File: Hunting Queries/TeamsLogs/NewExternalOrg.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  let data_date = 30d;
+  // If you want to look at users further back than the last day change this value
+  let lookback_data = 1d;
+  let known_orgs = (
+  TeamsData 
+  | where TimeGenerated > ago(data_date)
+  | where Operation == "MemberAdded" or Operation == "TeamsSessionStarted"
+  // Extract the correct UPN and parse our external organization domain
+  | extend UPN = iif(Operation == "MemberAdded", tostring(parse_json(Members)[0].UPN), UserId)
+  | extend Organization = tostring(split(split(UPN, "_")[1], "#")[0])
```
Comment (shainw) [inline Hunting Queries/TeamsLogs/NewExternalOrg.yaml]: We also have the format for UPN different than what you parse above for Organization - "user_contoso.com#EXT#@M365xXXXXX.onmicrosoft.com" that we may consider handling.  This format also comes thru as - user@M365xXXXXX.onmicrosoft.com.  Should we be handling this?

Comment (petebryan) [reply Hunting Queries/TeamsLogs/NewExternalOrg.yaml]: So we are only handling and parsing external users were, which all have the "user_contoso.com#EXT#@M365xXXXXX.onmicrosoft.com" format.
 user@M365xXXXXX.onmicrosoft.com are native tenant users so we are not interested in including them in this query.

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
+id: e9cef368-6257-4dab-9a48-4793729c46a2
+name: External user added and removed in short timeframe
+description: |
+  'This hunting query identifies external user accounts that are added to a Team and then removed within
+  one hour.
+  This query requires you to have Teams data collected from the O365 Management Activity API
```
[the remark this answers is not in the document]
Comment (petebryan) [reply Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: Thats the plan, once the blog is ready I will do another PR updating all of these.

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  | extend UPN = tostring(parse_json(Members)[0].UPN)
+  | project TimeAdded=TimeGenerated, Operation, UPN, UserWhoAdded = UserId
+  | join (
+  TeamsData 
+  | where TimeGenerated > ago(time_ago)
+  | where Operation == "MemberRemoved"
+  | extend UPN = tostring(parse_json(Members)[0].UPN)
+  | project TimeDeleted=TimeGenerated, Operation, UPN, UserWhoDeleted = UserId) on UPN
+  | where TimeDeleted < (TimeAdded + time_delta)
+  | project TimeAdded, TimeDeleted, UPN, UserWhoAdded, UserWhoDeleted
```
[the remark this answers is not in the document]
Comment (petebryan) [reply Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: Nice idea, I've done this for a few of the queries.

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml (hunk 2 above)
Comment (petebryan) [reply Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: Done

File: Hunting Queries/TeamsLogs/NewBotAdded.yaml (hunk 7 above)
Comment (petebryan) [reply Hunting Queries/TeamsLogs/NewBotAdded.yaml]: Fixed

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml (hunk 3 above)
Comment (petebryan) [reply Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: Done, and extracted TeamGuid in a few other places as well for clarity.

File: Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml (hunk 1 above)
Comment (petebryan) [reply Hunting Queries/TeamsLogs/ExternalUserAddedRemoved.yaml]: Can't be reliably done but will look to combine AAD and Teams data in a set of future hunting queries.

**Pull request merged**